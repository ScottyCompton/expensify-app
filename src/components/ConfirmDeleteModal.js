import React from 'react';
import Modal from 'react-modal';

const ConfirmDeleteModal = (props) => (
    <Modal
        isOpen={props.isOpen || false}
        contentLabel="Confirm Delete"
        onRequestClose={props.handleCloseModal}
        closeTimeoutMS={200}
        className="modal"
    >
        <p className="modal__title">Delete This Expense?</p>
        <button className="button btnCancel button--withspace button--secondary" onClick={props.handleCloseModal}>Cancel</button>
        <button className="button btnConfirm button--withspace button--warning" onClick={props.handleRemove}>Confirm</button>
    </Modal>
);

export default ConfirmDeleteModal;