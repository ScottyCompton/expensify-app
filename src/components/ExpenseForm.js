import expenses from "../reducers/expenses";
import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import { Link }  from 'react-router-dom';




export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : "",
            note: props.expense ? props.expense.note : "",
            amount: props.expense ? ((props.expense.amount)/100).toFixed(2) : "",
            dueDate: props.expense ? moment(props.expense.dueDate) : moment(),
            calendarFocused: false,
            error: "",
            btnText: props.expense ? "Update Expense" : "Add Expense"
        };
    }




    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        const re = /^\d{1,}(\.\d{0,2})?$/; 
        if(!amount || amount.match(re)) {
            this.setState(() => ({amount}))
        }
    };

    handleDateChange = (dueDate) => {
        if(dueDate) {
            this.setState(() => ({ dueDate }));
        }
        
    };


    onCalendarFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused}));
    };

    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount) {
            // set error state - please provide a description and an amount
            this.setState(() => ({error: "Please provide a desscription and an amount."}));
        } else {
            this.setState(() => ({error: ""}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                dueDate: this.state.dueDate.valueOf(),
                note: this.state.note
            })
        }

    };

    render () {
        return (

                <form className="form" onSubmit={this.onSubmit}>
                    {this.state.error && <p className="form__error">{this.state.error}</p>}
                    <label htmlFor="expense-description">Description:</label>
                    <input type="text"
                        id="expense-description" 
                        placeholder="Enter a brief description"
                        value={this.state.description}
                        className="text-input"
                        onChange={this.onDescriptionChange}
                        autoFocus 
                    />
                    <label htmlFor="expense-amount">Amount ($):</label>
                    <input type="number"
                        id="expense-amount"
                        placeholder="Enter the amount in USD ($)" 
                        className="text-input"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <label htmlFor="expense-duedate">Due Date:</label>
                    <SingleDatePicker
                        id={"expense-duedate"}
                        date={this.state.dueDate}
                        onDateChange={this.handleDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onCalendarFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <label htmlFor="expense-note">Expense Note (optional):</label>
                    <textarea 
                        placeholder="Add a note for this expense"
                        onChange={this.onNoteChange}
                        value={this.state.note}
                        className="textarea"
                        >
                    </textarea>
                    <div className="form__buttons">                    
                        <Link to="/dashboard" className="button button--secondary button--withleftspace">Cancel</Link>
                        {this.props.expense && <button className="button btnRemove button--secondary button--withleftspace button--secondary" type="button" onClick={this.props.handleOpenModal}>Remove</button>}
                        <button className="button btnSubmit button--withleftspace button--primary" type="submit">{this.state.btnText}</button>
                    </div>
                </form>
        );
    }
}

