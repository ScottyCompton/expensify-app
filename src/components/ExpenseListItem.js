import React from 'react';
import { Link }  from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ( {id, description, amount, dueDate} ) =>  (

        <Link className="list-item" to={`/edit/${id}`} title="Click to edit this expense">
          <div className="list-item__wrapper">
            <div className="list-item__title">
              <span>{description}</span>
            </div>
            <div className="list-item__duedate">
            <span className="show-for-desktop">{moment(dueDate).format('MMMM Do, YYYY')}</span>
            <span className="show-for-mobile">{moment(dueDate).format('M/D/YYYY')}</span>
            </div>
            <div className="list-item__data">
              <span>{numeral(amount/100).format('$0,0.00')}</span>
            </div>                  
          </div>
        </Link>


);

export default ExpenseListItem;