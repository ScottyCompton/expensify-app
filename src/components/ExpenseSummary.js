import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral'
import selectExpenseTotal from '../selectors/expense-total';
import selecteExpenses from '../selectors/expenses';

export const ExpenseSummary = (props) => {
    const totalRecs = props.expenses ? props.expenses.length : 0;
    const totalAmt = selectExpenseTotal(props.expenses);
    return (
        <div>Viewing {totalRecs} expense{(totalRecs >= 2 || totalRecs === 0) && <span>s</span>} totalling {numeral(totalAmt/100).format('$0,0.00')}</div>
    )
    
};

const mapStateToProps = (state) => {
    return {
        expenses: selecteExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseSummary);