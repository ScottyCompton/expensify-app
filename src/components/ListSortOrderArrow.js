import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import {Link} from 'react-router-dom';
import {faSortUp, faSortDown, faCaretRight} from '@fortawesome/free-solid-svg-icons'

const ListSortOrderArrow = (props) => {
    const inactiveArrow = (
        <Link to="/" onClick={props.handleClick}>
        <FontAwesomeIcon size="lg" icon={faCaretRight} />
        </Link>
    );

    const ascArrow = (
        <Link to="/" onClick={props.handleClick}>
        <FontAwesomeIcon size="lg" icon={faSortUp} />
        </Link>
    );

    const descArrow = (
        <Link to="/" onClick={props.handleClick}>
        <FontAwesomeIcon size="lg" icon={faSortDown} />
        </Link>
    );


    return (
        <span className={`sortorder-arrow sortorder-arrow--${props.sortOrder.toLowerCase()}`}>
            {props.sortOrder === 'INACTIVE' ? inactiveArrow : props.sortOrder === 'ASC' ? ascArrow : descArrow}
        </span>
    )
}

export default ListSortOrderArrow;