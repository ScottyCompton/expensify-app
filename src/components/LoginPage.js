import React from 'react'
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({startLogin}) => {
    return (
        <div>
            <div className="loginForm">
            <button onClick={startLogin} type="submit">Login</button>
            </div>
        </div>
    )
};

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);

