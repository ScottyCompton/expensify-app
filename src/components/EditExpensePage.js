import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import ConfirmDeleteModal from './ConfirmDeleteModal';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import { handleOpenModal, handleCloseModal } from '../actions/modal';

export class EditExpensePage extends React.Component {
    constructor(props) {
        super(props);
    }


    handleUpdate = (expense) => {
       this.props.startEditExpense(this.props.expense.id, expense);
       this.props.history.push('/');
    }

    handleRemove = () => {
        this.props.startRemoveExpense({id: this.props.expense.id});
        this.props.history.push('/');
        this.props.handleCloseModal();
    }



    render() {
        return (
        <div>
        
            
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Edit Expense</h1>
                </div>
            </div>

            <div className="content-container">
                <ExpenseForm 
                    onSubmit={this.handleUpdate}
                    expense={this.props.expense}
                    handleOpenModal={this.props.handleOpenModal} />
            </div>
            {<ConfirmDeleteModal 
            isOpen={this.props.isOpen}
            handleCloseModal={this.props.handleCloseModal}
            handleRemove={this.handleRemove}
            />}

        </div>
        );       
    }
}


const mapDispatchToProps = (dispatch) => {
    return {
        startEditExpense: (id, expense) => {dispatch(startEditExpense(id, expense))},
        startRemoveExpense: (id) => {dispatch(startRemoveExpense(id))},
        handleOpenModal: () => {dispatch(handleOpenModal())},
        handleCloseModal: () => {dispatch(handleCloseModal())},
    }
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id),
        isOpen: state.modal.isOpen
    }
};


export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
