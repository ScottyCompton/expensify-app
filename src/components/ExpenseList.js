import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selecteExpenses from '../selectors/expenses';
import ListSortOrderArrow from './ListSortOrderArrow';
import { setListDateSortOrder, setListAmountSortOrder } from '../actions/listSortOrder';
import { sortByDate, sortByAmount } from '../actions/filters';

export class ExpenseList extends React.Component {
    constructor(props) {
        super(props);
    }

    

    handleToggleDateSortOrder = (e) => {
        e.preventDefault();
        const sortOrder = this.props.listSortOrder.dateSortOrder === 'INACTIVE' ? 'ASC' :  this.props.listSortOrder.dateSortOrder === 'ASC' ? 'DESC' : 'ASC';
        this.props.toggleListDateSortOrder(sortOrder);
        this.props.sortByDate();
    }

    handleToggleAmountSortOrder = (e) => {
        e.preventDefault();
        const sortOrder = this.props.listSortOrder.amountSortOrder === 'INACTIVE' ? 'ASC' :  this.props.listSortOrder.amountSortOrder === 'ASC' ? 'DESC' : 'ASC';
        this.props.toggleListAmountSortOrder(sortOrder)
        this.props.sortByAmount();
    }
    
    render() {
        return (
            <div className="content-container">
                <div className="list-header">
                    <div className="show-for-mobile list-header-title">Expense</div>
                    <div className="show-for-desktop list-header-title">Expense</div>
                    <div className="list-header-duedate">
                        Due <span className="show-for-desktop"> Date</span>
                        <ListSortOrderArrow handleClick={this.handleToggleDateSortOrder} sortOrder={this.props.listSortOrder.dateSortOrder} />    
                    </div>
                    <div className="list-header-amount">
                        Amt
                        <ListSortOrderArrow handleClick={this.handleToggleAmountSortOrder} sortOrder={this.props.listSortOrder.amountSortOrder} />    
                    </div>
                </div>
                <div className="list-body">
                    {
                        this.props.expenses.length === 0 ? (
                            <div className="list-item list-item--message">
                                <span>No Expenses</span>
                            </div>
                            
                        ) : (
                            this.props.expenses.map((expense) => (
                                <ExpenseListItem key={expense.id} {...expense} />
                                )
                            )
                        )
                    }
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        expenses: selecteExpenses(state.expenses, {...state.filters, ...state.listSortOrder}),
        listSortOrder: state.listSortOrder
    };
}

const mapDispatchToProps = (dispatch) => {
    return {
        toggleListDateSortOrder: (sortOrder)  => dispatch(setListDateSortOrder(sortOrder)),
        toggleListAmountSortOrder: (sortOrder)  => dispatch(setListAmountSortOrder(sortOrder)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount())
    }

}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseList);
