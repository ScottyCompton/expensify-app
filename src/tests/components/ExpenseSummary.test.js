import React from 'react';
import { shallow } from 'enzyme';
import testData from '../fixtures/expenses';
import { ExpenseSummary } from '../../components/ExpenseSummary';


let expenses, wrapper;

beforeEach(() => {
    expenses = testData;
    wrapper = shallow(<ExpenseSummary expenses={expenses} />)
})


test('Should render ExpenseSummary Component', () => {
    expect(wrapper).toMatchSnapshot();
});


test('Should render total from a single entry', () => {
    const expense = expenses[0];

    wrapper.setProps({expenses: [expense]});
    expect(wrapper).toMatchSnapshot();
});


test('Should render with an empty array', () => {
    wrapper.setProps({expenses: []});
    expect(wrapper).toMatchSnapshot();
});
