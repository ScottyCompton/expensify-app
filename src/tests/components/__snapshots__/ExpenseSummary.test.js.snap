// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseSummary Component 1`] = `
<div>
  Viewing 
  3
   expense
  <span>
    s
  </span>
   totalling 
  $1,140.02
</div>
`;

exports[`Should render total from a single entry 1`] = `
<div>
  Viewing 
  1
   expense
   totalling 
  $0.02
</div>
`;

exports[`Should render with an empty array 1`] = `
<div>
  Viewing 
  0
   expense
  <span>
    s
  </span>
   totalling 
  $0.00
</div>
`;
